### PR TITLE
Change auth durability to store hash algorithm

### PR DESCRIPTION
**Pull request by Ignition**
To facilitate replication of auth information the hash algorithm used for the hash can not be just flag controlled --password-encryption-algorithm. 

Auth durability migration will introduce a new `version` key.
New "V1" will take the password_hash strings and make them a object which includes both algorithm and hashed password.

Also allow SHA256 hash algorithm to upgrade itself to include a hash once user has validated there password.

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [ ] Write a release note here, including added/changed clauses
- [ ] Tag someone from docs team in the comments

Changelog message: Added the password hashing algorithm into auth durability along side the password hash. Also sha256 based passwords hash will upon validation be upgraded to be rehashed with salt.

### CODE REVIEW

File: src/auth/models.cpp [hunk 1]
```diff
@@ -688,9 +691,17 @@ User User::Deserialize(const nlohmann::json &data) {
   if (!data.is_object()) {
     throw AuthException("Couldn't load user data!");
   }
-  if (!data["username"].is_string() || !data["password_hash"].is_string() || !data["permissions"].is_object()) {
+  auto password_hash_json = data["password_hash"];
+  if (!data["username"].is_string() || !(password_hash_json.is_object() || password_hash_json.is_null()) ||
```
Comment (andrejtonev) [inline src/auth/models.cpp:695]: What happens if the password_hash is nullptr?

File: tests/unit/auth.cpp [hunk 2]
```diff
@@ -987,25 +986,15 @@ TEST_F(AuthWithStorageWithVariousEncryptionAlgorithms, AddUserDefault) {
...
-
-TEST_F(AuthWithStorageWithVariousEncryptionAlgorithms, AddUserThrow) {
-  FLAGS_password_encryption_algorithm = "abcd";
-  ASSERT_THROW(auth.AddUser("Alice", "alice"), AuthException);
-}
-
-TEST_F(AuthWithStorageWithVariousEncryptionAlgorithms, AddUserEmptyPasswordEncryptionThrow) {
-  FLAGS_password_encryption_algorithm = "";
-  ASSERT_THROW(auth.AddUser("Alice", "alice"), AuthException);
-}
```
Comment (andrejtonev) [inline tests/unit/auth.cpp]: Why were these removed?

File: src/auth/auth.cpp [hunk 3]
```diff
@@ -64,7 +66,56 @@ const std::string kLinkPrefix = "link:";
...
+          puts.emplace(key, std::move(user_data));
+        } catch (const nlohmann::json::parse_error &e) {
+          throw AuthException("Couldn't load user data!");
+        }
+      }
+    }
+
+    // Perform migration to V1
+    store.PutMultiple(puts);
+    version_str = "V1";
```
[the remark this answers is not in the document]
Comment (Ignition) [reply src/auth/auth.cpp:111]: ATM it is a dead write, but sets up for the next version migration.

File: src/auth/crypto.cpp [hunk 4]
```diff
@@ -143,42 +147,70 @@ bool VerifyPassword(const std::string &password, const std::string &hash, const
...
+    throw AuthException("Invalid password encryption '{}'!", algo);
   }
+  return *maybe_parsed;
+}
 
-  throw AuthException("Invalid password encryption flag '{}'!", FLAGS_password_encryption_algorithm);
+PasswordHashAlgorithm &InternalCurrentEncryptionAlgorithm() {
+  static auto current = PasswordHashAlgorithm::BCRYPT;
+  static std::once_flag flag;
+  std::call_once(flag, [] { current = InternalParseEncryptionAlgorithm(FLAGS_password_encryption_algorithm); });
```
[the remark this answers is not in the document]
Comment (Ignition) [reply src/auth/crypto.cpp:178]: As my mind is right now, I think call_once is more intention revealing.

**Comment by Ignition** [comment]
related to #1359
